README.md
[![Build Status](https://travis-ci.org/tacnoman/dephpugger.svg?branch=master)](https://travis-ci.org/tacnoman/dephpugger) [![Code Climate](https://codeclimate.com/github/tacnoman/dephpug/badges/gpa.svg)](https://codeclimate.com/github/tacnoman/dephpug) [![HitCount](http://hits.dwyl.io/tacnoman/dephpugger.svg)](http://hits.dwyl.io/tacnoman/dephpugger) [![Coverage](https://codecov.io/gh/tacnoman/dephpugger/branch/master/graph/badge.svg)](https://codecov.io/gh/tacnoman/dephpugger) [![GitHub license](https://img.shields.io/github/license/tacnoman/dephpugger.svg)](https://github.com/tacnoman/dephpugger/blob/master/LICENSE) [![GitHub tag](https://img.shields.io/github/tag/tacnoman/dephpugger.svg)](https://GitHub.com/tacnoman/dephpugger/tags/) [![Packagist](https://img.shields.io/packagist/dt/tacnoman/dephpugger)](https://packagist.org/packages/tacnoman/dephpugger)


<img src="https://raw.githubusercontent.com/tacnoman/dephpugger/master/images/logo.png" alt="logo" title="Dephpugger logo" height="500">

# What is Dephpugger?

Dephpugger (read depugger) is an open source lib to make a debug in php direct in terminal, without necessary configure an IDE. The dephpugger run in Php Built in Server using another command. You can use for:

## Web applications
### Lumen in example
![dephpugger web](https://raw.githubusercontent.com/tacnoman/dephpugger/master/images/dephpugger-web.gif)
`Image 1.0 - Screenrecord for debug web`

## Cli Scripts
![dephpugger](https://raw.githubusercontent.com/tacnoman/dephpugger/master/images/dephpugger.gif)
`Image 1.1 - Screenrecord for debug cli scripts`

## Another example
![dephpugger](https://raw.githubusercontent.com/tacnoman/dephpugger/master/images/demo.gif)

# Install

To install you must run this code (using the composer).

```sh
$ composer require tacnoman/dephpugger
```

## Install globally
### In Linux or Mac Os X

Run this command:

```sh
$ composer global require tacnoman/dephpugger
```

And add in your ~/.bash_profile.

```
export PATH=$PATH:$HOME/.config/composer/vendor/bin
```

Now, run `source ~/.bash_profile` and you can run the commands using only `$ dephpugger`.

### In Windows

Sorry for this :/
I don't have Windos SO to run the Dephpugger, but I'm waiting your Pull Request to solve this ;D

But you can use Docker

### Using phar file

```bash
$ wget https://raw.githubusercontent.com/tacnoman/dephpugger/master/dephpugger.phar -o dephpugger
$ chmod +x ./dephpugger
$ mv ./dephpugger /usr/local/bin/dephpugger
```

Or installing via curl

```
$ curl https://raw.githubusercontent.com/tacnoman/dephpugger/master/installer.sh | bash
```

### Install XDebug

To install XDebug in Linux you must run:

```bash
$ sudo apt-get install php-7.2-cli php-dev php-pear
$ sudo pecl install xdebug
```

Find the last line, similar to: `You should add "zend_extension=/usr/lib/php/20170718/xdebug.so" to php.ini` and
copy to your `php.ini` file. To get the `php.ini` file, you can run:

```bash
$ php --info | grep php.ini

Configuration File (php.ini) Path => /etc/php/7.2/cli
Loaded Configuration File => /etc/php/7.2/cli/php.ini
```

Copy the `zend_extension=/usr/lib/php/20170718/xdebug.so` to last line in `/etc/php/7.2/cli/php.ini`.

### Docker

```
$ docker run --rm -p 8888:8888 tacnoman/dephpugger:latest server
```

# Dependencies

- PHP 7.0 or more (not tested in older versions)
- Xdebug activate
- A Plugin for your browser (If you want to debug a web application)

## Plugins for
- [Chrome](https://chrome.google.com/webstore/detail/xdebug-helper/eadndfjplgieldjbigjakmdgkmoaaaoc)
- [Firefox](https://addons.mozilla.org/pt-br/firefox/addon/the-easiest-xdebug/)
- [Safari](https://github.com/benmatselby/xdebug-toggler)
- [Opera](https://addons.opera.com/addons/extensions/details/xdebug-launcher/?display=en)

You can run this commands to check your dependencies:

```sh
$ vendor/bin/dephpugger requirements
$ vendor/bin/dephpugger info # To get all values setted in xDebug

# Or in global

$ dephpugger requirements
$ dephpugger info
```

# Usage

To usage you must (after installation) run two binaries in `vendor/bin` folder.

```sh
$ php vendor/bin/dephpugger debugger # Debugger waiting debug
$ php vendor/bin/dephpugger server   # Server running in port 8888

# Or in global

$ dephpugger debugger
$ dephpugger server
```

You must run in two different tabs (in next version you'll can run in an uniq tab).
After run theese commands, you need to put the follow line in your code:

```php
<?php
# ...
xdebug_break(); # This code is a breakpoint like ipdb in Python and Byebug in Ruby
# ....
```

Now, you can open in your browser the page (localhost:8888/[yourPage.php]).
When you request this page your terminal will start in breakpoint (like the image 1.0).

To debugger a php script, you could run:
```sh
$ php vendor/bin/dephpugger cli myJob.php

# Or in global

$ dephpugger cli myJob.php
```
## Comands after run

When you stop in a breakpoint you can make theese commands:

| Command             | Alias | Explanation                                                          |
|---------------------|-------|----------------------------------------------------------------------|
| next                | n     | To run a step over in code                                           |
| step                | s     | To run a step into in code                                           |
| set \<cmd>:\<value> |       | Change verboseMode or lineOffset in runtime                          |
| continue            | c     | To continue script until found another breakpoint or finish the code |
| list                | l     | Show next lines in script                                            |
| list-previous       | lp    | Show previous lines in script                                        |
| help                | h     | Show help instructions                                               |
| $variable           |       | Get a value from a variable                                          |
| $variable = 33      |       | Set a variable                                                       |
| my_function()       |       | Call a function                                                      |
| dbgp(\<command\>)   |       | To run a command in dbgp                                             |
| quit                | q     | Exit the debugger                                                    |

# Configuration (is simple)

The Dephpugger project has default options like a port, host, socket port, etc. You can change this values adding a file `.dephpugger.yml` in root directory project.
You can create in your `.dephpugger.yml` file the configurations. Like this:

```yml
--- 
debugger: 
  host: mysocket.dev # default: 0.0.0.0
  port: 9002 # default: 9005
  lineOffset: 10 # default: 6
  verboseMode: false # default: false
  historyFile: ~/.dephpugger_history # default: .dephpugger_history
server:
  host: myproject.dev # default: 0.0.0.0
  port: 8080 # default: 8888
  path: ./public/ # default: null
  file: index.php # default: null
```

Theese values will replace the default configuration.

# Full documentation

To see the full documentation [click here](http://dephpugger.com).

# How to use with phpunit, behat, codeception and others

The documentation to use, [click here](http://dephpugger.com/Usage/Running_with_phpunit.html).

# Run tests

```sh
$ composer test
$ composer lint
```

# Bugs?
Send me an email or open an issue:

Renato Cassino - Tacnoman - \<renatocassino@gmail.com\>

[See our changelog](https://github.com/tacnoman/dephpugger/blob/master/CHANGELOG.md)

# Articles and Videos

1. [How debug Drupal applications using Dephpugger](https://medium.com/@mosesliao/debug-drupal-using-dephpugger-a0839f878b56)
2. [Dephpugger in Sinapore event](https://www.youtube.com/watch?v=jmMgZ3GqbcQ)
3. [How debug applications with Dephpugger](https://hackernoon.com/how-debug-php-applications-with-dephpugger-98cc234d917c)
4. [Debugging with Dephpugger](https://laptrinhx.com/how-debug-php-applications-with-dephpugger-3582641373/)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [1.5.2 - 2019-10-01]
### Fixed
- Fix color code with reserved words

## [1.5.1 - 2019-09-30]
### Added
- Add command to generate phar file

### Changed
- Update symfony/yaml
- Update symfony/console
- Update cboden/ratchet
- Update php_codesniffer

## [1.5.0 - 2019-01-20]
### Added
- Requirements checking if sockets are installed
- Tutorial to run Dephpugger with Docker

### Fixed
- Support to Dephpugger

## [1.4.0 - 2019-01-20]
### Added
- Command to fix lint using PSR-2
- Phar file generator

## [1.3.0 - 2019-01-13]
### Added
- Docker full support

## [1.2.3 - 2019-01-12]
### Fixed
- Another solution to auloader (recursive)

## [1.2.2 - 2018-05-29]
### Fixed
- Bug in autoloader files to run project

## [1.2.0 - 2018-05-16]
### Fixed
- Bug with read .dephpugger.yml
- Read autoload file bug in some cases

### Changed
- Read autoload file mode

## [1.1.2 - 2017-11-14]
### Changed
- Fix command execution

### Changed
- Remove monolog

## [1.1.1 - 2017-05-30]
### Changed
- Change codeception to phpunit

## [1.1 - 2017-04-29]
### Added
- Getting global variables in property get
- Tests

## [1.0.2-rc1 - 2017-04-26]
### Added
- Class to client for Dbgp
- Tests for dbgp Server and Client
- Command 'set' to change the config

### Changed
- Namespace for client and server Dbgp
- Removing phpPath and add PHP_BINARY

### Fixed
- Command cli printing content

## [1.0.1-alpha]
### Fixed
- Bug in autoload in global

## [1.0.0-alpha]
### Added
- New architecture
- Tests for commands
- Tests for MessageEvent
- One class for each command in debugger
- One class for each xml received in DBGP
- New flow execution
- Core with REPL
- Class with reflection mode to get all plugins (commands and parsers)
- Readline with history

### Changed
- Refactor client DBGp protocol
- Remove singleton in $config
- Output in a single class

## [0.5.1] - 2017-03-24
### Fixed
- Bug in autoload

## [0.5.0] - 2017-03-22
### Added
- Add command Help
- Add history to debugger
- Create a unique class to print with symfony color
- Command `list-previous` to show previous lines in file

## [0.4.1] - 2017-03-19
### Changed
- Attribute verboseMode now is in debugger, not in options
- Command l to list
- If send command emtpy, repeat the last command sent

### Fixed
- Bug in command requirements and info

## [0.4.0] - 2017-03-19
### Added
- If send empty string, repeat the last command in debugger

### Changed
- Using DIRECTORY_SEPARATOR instead of '/' (windows support)
- Refactor in DBGP server
- Separate responsability to parse messages in one class

### Fixed
- The code were using 127.0.0.1 in debugger server instead of variable setted

## [0.3.1] - 2017-03-11
### Added
- Sent messages in verboseMode

### Fixed
- Bug with set a variable

## [0.3.0] - 2017-03-11
### Added
- Support to show attributes in object var
- Test for string
- Get response when call a method

### Changed
- Refactor in FilePrinter to use xml instead of regex
- Use xml in tests

### Fixed
- Bug with objects debugger. Fixed when use var $this

## [0.2.1] - 2017-03-06
### Fixed
- Bug with path and without file

## [0.2.0] - 2017-03-06
### Added
- Params file and path

## [0.1.2] - 2017-03-01
### Added
- Catch error if try show value of unexist var

## [0.1.1] - 2017-03-01
### Added
- When show type array, show in php format, not xml
- When show type object, show the name of the class
- More tests for CommandAdapter
- Command to get info about xdebug
- Add option to offset lines in .dephpugger.yml

### Fixed
- Add type float to show variable in array

## [0.0.3] - 2017-03-01
### Added
- Add circleci support
- Refactor client DBGp protocol - [WIP]
- Class to get variable in phpinfo();
- Create exception to quit dephpugge
- Run dephpugger for php scripts in cli
- Counter changing the transactionId

## Changed
- Algorithm to get data from DBGPp protocol
- Quit command in class CommandAdapter

## Fixed
- Bug when stop a script
- Fix the stop when debugger start

## [0.0.2]
### Changed
- Use symfony/console for all commands

### Added
- Explanation how to use the .dephpugger.yml
- File .dephpugger.yml in root path to change default configs
- Add type to return format variable
- Add a (ridiculuous) logo to github
- Can run in verbose mode to print all xml DBGp
- Add command to check requirements

## [0.0.1]
### Added
- TravisCI integration
- Codeclimate integration
- Returning formated values instead of XML DBGp format
- Color code with symfony/console
- Figlet to write Dephpugger in terminal for server and debugger
- Create README explain the project with an image (screenshot)
- Install monolog to log application
- Set a current version in a separated file
- Create changelog
- Tests for commands and FilePrinter
- FilePrinter coloring code with reserved words
- FilePrinter class to print a file in terminal
- Command conversor to IDE commands
- Possibility to run IDE commands to DBGp
- Breakpoints
- Install codeception to make tests
- Binary file to run client websocket to DBGp protocol
- Binary file to run php server built in connecting websocket
- PSR-0 implemented
- Added to packagist
- Composer component to install
- Client to DBGp protocol


src/Dephpug/Config.php
<?php

namespace Dephpug;

use Symfony\Component\Yaml\Yaml;
use Symfony\Component\Yaml\Exception\ParseException;

/**
 * Class to get the default config using file *.dephpugger.yml*
 */
class Config
{
    /**
     * Default config that will be replaced by *.dephpugger.yml* file
     */
    private $defaultConfig = [
        'server' => [
            'port' => 8888,
            'host' => '0.0.0.0',
            'path' => null,
            'file' => '',
        ],
        'debugger' => [
            'port' => 9005,
            'host' => '0.0.0.0',
            'lineOffset' => 6,
            'verboseMode' => false,
            'historyFile' => '.dephpugger_history',
        ],
    ];

    /**
     * Array with configurations
     */
    private $config;

    /**
     * Get the attribute config
     *
     * @return array $config
     */
    public function getConfig()
    {
        return $this->config;
    }

    /**
     * Read the file (if exists) .dephpugger.yml
     */
    public function configure()
    {
        $config = $this->getConfigFromFile();
        $this->config = array_replace_recursive($this->defaultConfig, $config);
    }

    /**
     * Get path of file *.dephpugger.yml*
     */
    public function getPathFile()
    {
        return getcwd().'/.dephpugger.yml';
    }

    /**
     * Get configuration from file *.dephpugger.yml*[
     *
     * @return array $config
     */
    public function getConfigFromFile()
    {
        $config = [];
        $path = $this->getPathFile();
        if (file_exists($path)) {
            try {
                $config = Yaml::parse(file_get_contents($path));
            } catch (ParseException $e) {
                throw new ParseException('The file .dephpugger.yml is invalid');
            }
        }

        return $config;
    }

    /**
     * Magit Method to get config as an attribute
     *
     * @return array|null
     */
    public function __get($key)
    {
        if (isset($this->config[$key])) {
            return $this->config[$key];
        }

        return null;
    }

    /**
     * Change the attribute in configuration debugger
     *
     * @param  string $key   Indicates the debugger's attribute
     * @param  string $value Indicates the new value
     * @return void
     */
    public function setNewDebuggerValue($key, $value)
    {
        if (isset($this->config['debugger'][$key])) {
            $this->config['debugger'][$key] = $value;
        }
    }
}
